Wildcards nodes and units

engelmi opened the pull request:
This PR attempts to fix #185 by adding wildcard `*` support for nodes and units. 

In addition, it also contains:   
- Change to the `Subscribe` and `Unsubscribe` API to accept a list of node-unit-tuples
- Extending `hirtectl` by a (basic) implementation for creating a monitor + subscription(s)

engelmi committed:
moved string util functions to dedicated header

Signed-off-by: Michael Engel <mengel@redhat.com>

engelmi committed:
implemented wildcard and list support for node and unit subscriptions

Signed-off-by: Michael Engel <mengel@redhat.com>

engelmi committed:
added basic monitoring cmd to hirtectl

Signed-off-by: Michael Engel <mengel@redhat.com>

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1076,6 +1087,11 @@ static int agent_method_unsubscribe(sd_bus_message *m, void *userdata, UNUSED sd
         }
 
+        if (streq(unit, "*")) {
+                agent->wildcard_subscription_active = false;
```
Maybe check that wildcard_subscription_active is already set and error if not?

engelmi replied:
Yes, added that check.

alexlarsson commented on src/manager/monitor.c:
```diff
@@ -194,6 +195,84 @@ static int monitor_method_close(sd_bus_message *m, void *userdata, UNUSED sd_bus
+                        success = false;
+                        break;
+                }
+        }
```
There was already support in manager.c to handle node == NULL in subscribe meaning to match for all nodes. However, you removed that and replaced it with this loop instead. I don't really think this approach is better. You will just get more Subscription objects, and you need the cleanup unsubscribe below that wasn't otherwise needed.

Is there a pro to having one Subscription per node in this case?

engelmi replied:
Not really. I wanted to move that logic into the monitor, but - thinking about it - it doesn't really fit there. This part has changed with the latest changes based on our discussion and is now similar to the previous implementation.

alexlarsson commented on src/manager/monitor.c:
```diff
@@ -194,6 +195,84 @@ static int monitor_method_close(sd_bus_message *m, void *userdata, UNUSED sd_bus
+        }
+
+        char *saveptr = NULL;
+        char *unit_name = strtok_r(unit_copy, ",", &saveptr);
```
I don't like using stringly typed things like this. If we want to pass multiple units, we should do it with actually typed dbus arrays.

engelmi replied:
I totally agree. This has been removed in one of the subsequent commits. By the latest rebase I cleaned up that commit as well - its now in the `SubscribeList` API method.

alexlarsson commented on src/manager/monitor.c:
```diff
@@ -244,6 +248,45 @@ static int monitor_method_subscribe(sd_bus_message *m, void *userdata, UNUSED sd
+                if (is_wildcard(node)) {
+                        subscribe_on_all(manager, monitor, unit);
+                } else {
+                        subscribe_on_node(manager, monitor, node, unit);
```
I really would like a single Subscription objects with a list of units in it rather than multiple Subscriptions like this. Then Unsubscribe should just remove this one object.

alexlarsson commented on src/manager/monitor.c:
```diff
@@ -46,6 +46,38 @@ Subscription *create_monitor_subscription(Monitor *monitor, const char *node, co
+        return sub->node != NULL && streq(sub->node, SYMBOL_WILDCARD);
+}
+
+bool subscription_has_unit_wildcard(Subscription *sub) {
```
I would probably do the is_wildcard() check in add_unit() and just set an is_wildcard boolean which you return from here.

engelmi replied:
Good point, but I actually was able to remove it due to merging the wildcard list into the existing unit_subscription hashmap.

alexlarsson commented on src/manager/node.c:
```diff
@@ -174,6 +191,14 @@ void node_unref(Node *node) {
+        LIST_FOREACH_SAFE(wildcard_subscriptions, sub, next_sub, node->wildcard_subscriptions) {
+                manager_remove_subscription(node->manager, sub);
+                LIST_REMOVE(wildcard_subscriptions, node->wildcard_subscriptions, sub);
+                subscription_unref(sub);
```
This part doesn't look right to me. Yes, we should maybe remove the sub from wildcard_subscriptions, because node_subscribe_wildcard_unit() added it there. However, we should not unref it, because we didn't ref it.  Alterntively we should ref the sub when adding it, but I don't think that is right, as we risk creating cycles. And anyway the node will not be unref:ed until shutdown, so lifetimes are not really a problem.

And we shouldn't call manager_remove_subscription(), which will remove the sub from other nodes too. That should only happen when the subscription is unsubscribed, which comes from the monitor.

engelmi replied:
Removed this part completely since I merged the `wildcard_subscriptions` into the existing hashmap for the `unit_subscriptions`.

alexlarsson commented on src/manager/node.c:
```diff
@@ -799,31 +884,33 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
-                        int r = sub->handle_unit_removed(sub->monitor, node->name, sub->unit, "virtual");
-                        if (r < 0) {
-                                hirte_log_error("Failed to emit UnitRemoved signal");
+                                if (send_state_change) {
```
This changes the order of the unit removed and the state changed events.
Also, I think i had some reentrency issue in  the case where handle_unit_removed destroyed the monitor, which made me do the double LIST_FOREACH_SAFE loops that you merged here.

engelmi replied:
Ups, you are right - I didn't think about the order here. Reverted it.

alexlarsson commented on src/manager/node.c:
```diff
@@ -799,31 +884,33 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
 
                 usubs->loaded = false;
+
                 UnitSubscription *usub = NULL;
```
Hmm, this only sends the virtual events for non-wildcard subscriptions. Also, we don't even track the last state for such subscriptions, so we can't sent any state change event...

engelmi replied:
As we discussed, I changed this so that virtual `UnitNew` and `UnitRemoved` (also StateChanged) signals are emitted. This way the monitor gets notified if, for example, the node disconnects and reconnects, and can act upon it (e.g. query current state of special units).

alexlarsson commented on src/manager/node.c:
```diff
@@ -1397,105 +1484,154 @@ static void node_send_agent_subscribe_all(Node *node) {
+                /* We know this is loaded, so we won't get notified from
+                   the agent, instead send a virtual event here. */
+                if (usubs->loaded) {
+                        int r = sub->handle_unit_new(sub->monitor, node->name, sub_unit->name, "virtual");
```
Same here, none of these virtual events are sent for a wildcard subscription.

alexlarsson commented on src/manager/node.h:
```diff
@@ -50,6 +50,7 @@ struct Node {
         LIST_HEAD(ProxyMonitor, proxy_monitors);
         LIST_HEAD(ProxyDependency, proxy_dependencies);
 
+        LIST_HEAD(Subscription, wildcard_subscriptions);
```
Alternatively you could store the wildcard subscription in the unit_subscription hashmap under the "*" key. Not sure which is cleaner.

engelmi replied:
I tried that approach and I could simplify a bit of code - so I'd go with that.

alexlarsson commented on src/manager/node.c:
```diff
@@ -1397,105 +1484,154 @@ static void node_send_agent_subscribe_all(Node *node) {
+void node_unsubscribe_wildcard_unit(Node *node, Subscription *sub) {
+        LIST_REMOVE(wildcard_subscriptions, node->wildcard_subscriptions, sub);
+        if (!node_has_wildcard_subscriptions(node)) {
+                /* First wildcard sub to this node, pass to agent */
```
Should comment be "Last wildcard" maybe?

engelmi replied:
Yes, copy & paste error :sweat_smile:

alexlarsson commented on doc/dbus-interfaces.md:
```diff
@@ -76,16 +76,24 @@ Object path: `/org/containers/hirte/monitor/$id`
+    Subscribe to changes in properties of a given unit on a given node. Passing a wildcard `*` to either `node` or
+    `unit` will subscribe to all nodes or all units on the given node. This will emit the signal `UnitChanged` for all
+    matching units in the system, and then again whenever one of the properties of the unit changes. Returns an
+    identifier `id` used for a subsequent `Unsubscribe`.
```
This isn't needed in this PR, but at some point we need to extend this with a fuller description of the events, including the virtual events (and how they differ for wildcarded matches).

engelmi replied:
Good point, I created [an issue](https://github.com/containers/hirte/issues/261) to keep track of this task.

alexlarsson commented on src/agent/agent.c:
```diff
@@ -1043,6 +1044,19 @@ static int agent_method_subscribe(sd_bus_message *m, void *userdata, UNUSED sd_b
+                }
+                agent->wildcard_subscription_active = true;
+
+                AgentUnitInfo info = { NULL, (char *) unit, true, true, UNIT_ACTIVE, NULL };
```
UNIT_ACTIVE for a "*" unit seems a bit weird to me, but unfortunately I don't really see a better alternative, and I don't think we want to add new possible states.

alexlarsson replied:
I think we should have a substate though.

engelmi replied:
Yes, I agree that the `UnitNew` doesn't really fit. The unit new signal only contains the name of the unit and the reason - so every other information like the active state is not used. `AgentInfo` is only created for the sake of calling the `agent_emit_unit_new` function. Maybe replacing `UNIT_ACTIVE` with a -1 is more clear?

alexlarsson replied:
_UNIT_ACTIVE_STATE_INVALID maybe? Although I don't really think it matters. Lets leave it.

alexlarsson commented on src/manager/node.c:
```diff
@@ -260,53 +276,97 @@ static int node_match_unit_properties_changed(sd_bus_message *m, void *userdata,
+                if (is_wildcard(unit)) {
+                        usubs->active_state = UNIT_ACTIVE;
+                        free(usubs->substate);
+                        usubs->substate = strdup("running");
```
This initializes the virtual substate for a wildcard to "running", so maybe we should use that in the agent too?

engelmi replied:
Contrary to the manager, in the agent the wildcard is only represented as a boolean flag and is not traced in any other way. If active, the agent checks both, the subscription of the specific unit (just like before) and, in addition, the wildcard flag.

alexlarsson commented on src/manager/node.c:
```diff
@@ -260,53 +276,97 @@ static int node_match_unit_properties_changed(sd_bus_message *m, void *userdata,
+                        Subscription *sub = usub->sub;
+
+                        /* only emit signal once per monitor */
+                        if (usubs != NULL &&
```
This is called in several places, so maybe you can move the usubs != NULL check into unit_subscription_contains_monitor.

engelmi replied:
Done. But the `unit_subscription_contains_monitor` has been removed later in favor of building a unique monitor list.

alexlarsson commented on src/manager/node.c:
```diff
@@ -260,53 +276,97 @@ static int node_match_unit_properties_changed(sd_bus_message *m, void *userdata,
+                UnitSubscription *next_usub = NULL;
+                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
+                        Subscription *sub = usub->sub;
+                        int r = sub->handle_unit_property_changed(sub->monitor, node->name, unit, interface, m);
```
This is not your fault, but I think there is another issue here. What if a single monitor has multiple subscriptions that overlap. We will then send multiple events to this single monitor.

I wonder if we should have the two loops just compute a vector of of (unique) monitors, and then loop over those and call handle_unit_property_changed.

engelmi replied:
I moved the `handle_xyz` functions from the `Subscription` to the `Monitor`, which seems to be a better fit. When I tried to compute a unique list of monitors with only a hashmap (using the monitor id as key), that resulted in segfaults due to the `handle_xyz` functions being NULL - since the hashmap copies everything. I decided to use a hashmap just for the ids to track which monitor is already added to a list of unique monitors. 
This also results in the `unit_subscription_contains_monitor` to be removed.

alexlarsson commented on src/manager/node.c:
```diff
@@ -799,31 +897,45 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
-                                if (r < 0) {
-                                        hirte_log_error("Failed to emit UnitRemoved signal");
+
+                                SubscribedUnit *unit_sub = NULL;
```
This seems wrong to me. On subscribe we take the list of subscribed units in the user-specified subscription and for each of the specified units create a new UnitSuibscription and put in the hashtable. So, now that you interate over all the hashtable entries you should get every unit name ever requested. So, there is no need to now iterate over all the units in this subscription again.

In fact, it is very wrong. The send_state_change and the active_state computed above that we are sending *only* applies to the unit we're currently looking at from the hashtable.

engelmi replied:
Good catch! I remove the unnecessary loop.

alexlarsson commented on src/manager/node.c:
```diff
@@ -799,31 +897,45 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
-                                hirte_log_error("Failed to emit UnitRemoved signal");
+                        SubscribedUnit *unit_sub = NULL;
+                        SubscribedUnit *next_unit_sub = NULL;
+                        LIST_FOREACH_SAFE(units, unit_sub, next_unit_sub, sub->subscribed_units) {
```
same here

alexlarsson commented on src/manager/node.c:
```diff
@@ -1394,104 +1506,112 @@ static void node_send_agent_subscribe_all(Node *node) {
+                /* We know this is loaded, so we won't get notified from
+                   the agent, instead send a virtual event here. */
+                if (usubs->loaded) {
+                        int r = sub->handle_unit_new(sub->monitor, node->name, sub_unit->name, "virtual");
```
This is also potentially sending the same "new" unit twice if you subscribe to the same unit twice. I wonder if we should track that we sent it? This is kinda fringe semantics though.

alexlarsson commented on src/manager/node.c:
```diff
@@ -262,18 +246,38 @@ static int node_match_job_state_changed(
+                        monitor_hashmap_compare,
+                        NULL,
+                        NULL);
+        if (unique_monitor_ids == NULL) {
```
This use of hash is rather weird. What you want to do is put the actual monitor pointers into the hash, and then just return the resulting hashtable and have the caller use hashmap_iter() to loop over the monitors.

You need:

```
static uint64_t monitor_hashmap_hash(const void *item, uint64_t seed0, uint64_t seed1) {
   return (uint64_t)item;
}

static int monitor_hashmap_compare(const void *a, const void *b, UNUSED void *udata) {
   return a == b;
}
```

Then pass size as sizeof (void *), and on insert:

```
hashmap_set(unique_monitors, mo); /* Ignore result, no need to free old value */
if (hashmap_oom(unique_monitor_ids)) {
 ...
}
```

alexlarsson replied:
Ah, actually, compare should be if (a == b) return 0; else return 1;

alexlarsson commented on src/manager/node.c:
```diff
@@ -334,38 +380,18 @@ static int node_match_unit_new(sd_bus_message *m, void *userdata, UNUSED sd_bus_
+                        monitors,
+                        unique_monitor,
+                        unique_monitor_next,
+                        node_compute_unique_monitor_list(node, unit)) {
```
Is this really right if `unit == "*"`? I'm not sure. Seems like we may sometimes send both virual new "*" from the manager and then the agent in some cases. Needs more careful analysis.

engelmi replied:
I think this should be alright for `*`.  The agent sends only a virtual UnitNew for wildcards when such a subscription is made - which is forwarded by the manager to the monitor. Since wildcard units and "regular" units are stored in the `unit_subscriptions` of the respective node, `node_compute_unique_monitor_list` prevents duplicate emitting the UnitNew signal to the monitor. 
If the node disconnects, then the monitor with the wildcard gets treated the same as regular unit subs and a StateChanged (with the `agent-offline`) substate as well as a UnitRemoved gets emitted for it. On a reconnect of the agent, the manager subscribes with the wildcard again and we receive another UnitNew signal. 
Here is a sample output for subscribing, disconnecting the agent and reconnecting again:
```
Monitor path: /org/containers/hirte/monitor/5
Subscribing to node 'laptop' and unit '*'
[laptop] *
        Unit created (reason: virtual)
[laptop] *
        Unit state changed (reason: virtual)
        Active: inactive (agent-offline)
[laptop] *
        Unit removed (reason: virtual)
[laptop] *
        Unit created (reason: virtual)

```

alexlarsson commented on src/manager/node.c:
```diff
@@ -901,38 +891,27 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
 
                 if (send_state_change) {
                         LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
```
I wonder if this should not also call node_compute_unique_monitor_list() for name=usubs->name? I can imagine multiple subscriptions for the same monitor?

I guess that would behave somewhat different, as it would also send events on disconnect for actual units to wildcarded monitors. Maybe node_compute_unique_monitor_list() can have a bool include_wildcards argument?

engelmi replied:
Yes, good point. I'll change it to use the `node_compute_unique_monitor_list` instead. 
I think we want to have the same behavior for wildcard subs as we have for regular ones. This ensures that the user of the monitor knows when an agent disconnected and it might want to catch up with the latest state if it reconnects.

alexlarsson commented on src/manager/monitor.c:
```diff
@@ -37,11 +37,6 @@ Subscription *create_monitor_subscription(Monitor *monitor, const char *node) {
         subscription->monitor = monitor_ref(monitor);
         subscription->free_monitor = (free_func_t) monitor_unref;
 
-        subscription->handle_unit_new = monitor_on_unit_new;
```
Ideally you should break out this change to a separate commit.

engelmi replied:
Done.

engelmi committed:
moved handle signal functions from Subscription to Monitor

Signed-off-by: Michael Engel <mengel@redhat.com>

alexlarsson commented on src/manager/node.c:
```diff
@@ -791,31 +870,39 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
                         }
-                }
 
-                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
```
I think you need to call node_compute_unique_monitor_list() again here, because the call to handle_unit_state_changed() above may have removed the monitor.

alexlarsson committed:
manager: Fix manager_remove_subscription() of wildcard subscriptions

We need to remove from all_subscriptions and unref, even for
wildcards subscriptions.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson committed:
manager: Handle shutdown better

When finalizing manager, remove subscriptions properly such
that the nodes get told of the subscriptions going away.

Also, free the nodes after we remove the subscriptions so that
they are properly alive when this happens.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson reviewed (approved):
LGTM
